packages/@aws-cdk/aws-s3-deployment/lambda/src/index.py

```diff
@@ -29,41 +31,64 @@ def cfn_error(message=None):
         # extract resource properties
         props = event['ResourceProperties']
         old_props = event.get('OldResourceProperties', {})
+        physical_id = event.get('PhysicalResourceId', None)
 
         try:
             source_bucket_name = props['SourceBucketName']
             source_object_key  = props['SourceObjectKey']
             dest_bucket_name   = props['DestinationBucketName']
             dest_bucket_prefix = props.get('DestinationBucketKeyPrefix', '')
-            retain_on_delete   = props.get('RetainOnDelete', "false") == "true"
+            retain_on_delete   = props.get('RetainOnDelete', "true") == "true"
         except KeyError as e:
-            cfn_error("missing request resource property %s" % str(e))
+            cfn_error("missing request resource property %s. props: %s" % (str(e), props))
             return
 
+        # treat "/" as if no prefix was specified
+        if dest_bucket_prefix == "/":
+            dest_bucket_prefix = ""
+
         s3_source_zip = "s3://%s/%s" % (source_bucket_name, source_object_key)
         s3_dest = "s3://%s/%s" % (dest_bucket_name, dest_bucket_prefix)
 
         old_s3_dest = "s3://%s/%s" % (old_props.get("DestinationBucketName", ""), old_props.get("DestinationBucketKeyPrefix", ""))
+
+        # obviously this is not
+        if old_s3_dest == "s3:///":
+            old_s3_dest = None
+
         logger.info("| s3_dest: %s" % s3_dest)
         logger.info("| old_s3_dest: %s" % old_s3_dest)
 
+        # if we are creating a new resource, allocate a physical id for it
+        # otherwise, we expect physical id to be relayed by cloudformation
+        if request_type == "Create":
+            physical_id = "aws.cdk.s3deployment.%s" % str(uuid4())
```

Comment:
Why not just use the target S3 bucket location as physical ID?

Reply:
The destination is mutable and I have code that handles updates for it by first deleting the old and then uploading to new (and not the opposite). This is to address the situation where for example old destination is a parent directory of the new destination